src/passes/OptimizeInstructions.cpp

```diff
@@ -2023,97 +2061,61 @@ struct OptimizeInstructions
       //
       // As above, we must refinalize as we may now be emitting a more refined
       // type (specifically a more refined heap type).
-      replaceCurrent(Builder(*getModule()).makeRefAs(RefAsNonNull, curr->ref));
+      replaceCurrent(builder.makeRefAs(RefAsNonNull, curr->ref));
       refinalize = true;
       return;
     }
 
-    // Repeated identical ref.cast operations are unnecessary. First, find the
-    // immediate child cast, if there is one.
-    // TODO: Look even further through incompatible casts?
-    auto* ref = curr->ref;
-    while (!ref->is<RefCast>()) {
-      auto* last = ref;
-      // RefCast falls through the value, so instead of calling
-      // getFallthrough() to look through all fallthroughs, we must iterate
-      // manually. Keep going until we reach either the end of things
-      // falling-through, or a cast.
-      ref = Properties::getImmediateFallthrough(ref, passOptions, *getModule());
-      if (ref == last) {
-        break;
-      }
-    }
-    if (auto* child = ref->dynCast<RefCast>()) {
+    if (auto* child = curr->ref->dynCast<RefCast>()) {
       // Repeated casts can be removed, leaving just the most demanding of
-      // them.
+      // them. Note that earlier we already checked for the cast of the ref's
+      // type being more refined, so all we need to handle is the opposite, that
+      // is, something like this:
+      //
+      //   (ref.cast $B
+      //     (ref.cast $A
+      //
+      // where $B is a subtype of $A. We don't need to cast to $A here; we can
+      // just cast all the way to $B immediately.
+      if (Type::isSubType(curr->type, child->type)) {
+        curr->ref = child->ref;
+        return;
+      }
+
+      // As above, we can also consider the case where the heap type of the
+      // child is a supertype even if the type as a whole is not, which means
+      // that nullability is an issue, specifically in the form of the child
+      // having a heap supertype which is non-nullable, and the parent having
+      // a heap subtype which is nullable, like this:
+      //
+      //   (ref.cast null $B
+      //     (ref.cast $A
+      //
+      // We can optimize that to
+      //
+      //   (ref.cast $B
+      //     (ref.as_non_null $A
```

Comment:
Would it be worth saving the allocation of the `ref.as_non_null` node and finishing here instead of falling through to the next rule?

Reply:
Yeah, I guess it's almost as simple to just optimize it fully there, I'll do that.